File: kotlin/raindrops/src/main/kotlin/Raindrops.kt
object Raindrops {
    fun convert(input: Int): String {
        val rules = sequenceOf(
                3 to "Pling",
                5 to "Plang"
        )

        val result = rules
                .filter { (denominator, _) -> input.isDivisibleBy(denominator) }
                .map { (_, result) -> result }
                .joinToString("")
        return if (result.isNotEmpty()) result else input.toString()
    }

    private fun Int.isDivisibleBy(denominator: Int) = this % denominator == 0

}

File: typescript/raindrops/raindrops.ts
export default class Raindrops {

    private rules = [
        {appliesTo: this.isDivisibleBy(3), result: 'Pling'},
        {appliesTo: this.isDivisibleBy(5), result: 'Plang'},
        {appliesTo: this.isDivisibleBy(7), result: 'Plong'},
    ]

    convert(input: number) {
        const match = this.rules
            .filter((r) => r.appliesTo(input))
            .map((r) => r.result)
            .join('')
        return match || '1'
    }

    private isDivisibleBy(divisor: number) {
        return (input: number) => {
            return input % divisor === 0
        }
    }
}